### PR TITLE
create bitcoin.conf on first run with template

### LINKED ISSUE #10746: Create a default bitcoin.conf in the correct location.
A recurring problem when helping new users getting started with running Bitcoin Core is the config file location; a lot of uncertainty is caused by the config file and the folders it should go in not being created automatically by the installation process.

It does not even have to be the full sample as found in contrib, just having an empty config file (or one with a minimal comment directing people to the sample file) created by default would ease the assisting of users on OSes one is not familiar with, as it would be easier for them to verify that they are putting their OS-agnostic configuration in the correct OS-specific location without some unnecessary trial-and-error.

### PR DESCRIPTION
**Pull request by ghost**
Fixes https://github.com/bitcoin/bitcoin/issues/10746 and https://github.com/bitcoin/bitcoin/issues/19641

### What?

Create bitcoin.conf on first run with a template added as comments. Can add/remove things based on feedback.

### Why?

`bitcoind` can already create bitcoin.conf if a user runs below command on first startup:

```
bitcoind -startupnotify="touch path/bitcoin.conf"
```

`bitcoin-qt` can already create blank bitcoin.conf if a user can find this option after starting Bitcoin Core for the first time:

image

This pull request improves UX when using `bitcoind`.

### How?

Used most of the code from https://github.com/bitcoin/bitcoin/pull/13761/ (will add Alexander Leishman as co-author if I could get the email address)

Template is changed and I have used bitcoin.conf (privacy) template from https://github.com/jlopp/bitcoin-core-config-generator/

### CODE REVIEW

**Comment by prayank23** [comment]
I tried compiling and running this locally. Weird this had no errors when compiling on Fedora but there was no bitcoin.conf created on running bitcoind.

I will try testing on other OS tomorrow.

**Comment by meshcollider** [comment]
> will add Alexander Leishman as co-author if I could get the email address

You can fetch his branch from the PR by running
```git fetch upstream pull/13761/head```
And then using `git show` to look at one of the commits and see the email address used.

File: src/conf_file.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# An option can be confined to one network by adding in the relevant section or adding network before option. Example: test.rpcport = 18332
+# EXCEPTIONS: The options addnode, connect, port, bind, rpcport, rpcbind and wallet
+# only apply to mainnet unless they appear in the appropriate section below.
+
+# Options only for mainnet
+#[main]
+
+# Options only for testnet
+#[test]
+
```
Comment (benthecarman) [inline src/conf_file.h:46]: nit: should add a signet section

File: src/util/system.cpp [hunk 2]
```diff
@@ -447,14 +448,24 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
         path /= fs::PathFromString(BaseParams().DataDir());
 
     if (fs::create_directories(path)) {
-        // This is the first run, create wallets subdirectory too
-        fs::create_directories(path / "wallets");
+        // This is the first run, create wallets subdirectory and bitcoin.conf
```
Comment (benthecarman) [inline src/util/system.cpp:451]: this comment should probably go in the docs for `SetupDatadir`

Comment (prayank23) [reply src/util/system.cpp:451]: Not sure about this. This comment was already present. I just added `and bitcoin.conf` in it.

File: src/conf_file.h (hunk 1 above)
Comment (prayank23) [reply src/conf_file.h:46]: Done. Added in https://github.com/bitcoin/bitcoin/pull/23931/commits/5dabe0a0379cdf35faeab72a3cd7f65bc6d930c6

**Comment by prayank23** [comment]
Thanks @meshcollider I was able to find the email address and added as co-author.

Changes in last commit:

1. Add Alexander Leishman as co-author  

2. Add signet section

3. Change copyright year

4. Change everything after `#endif` to comments based on errors in CI

Tried testing and it works as expected. Config file was created on first run. It works when there is no data directory present in default location which was not the case when I tried earlier in https://github.com/bitcoin/bitcoin/pull/23931#issuecomment-1003584246

**Comment by prayank23** [comment]
I have marked the pull request ready for review. Will figure out a way to add test if it is required.

**Comment by Rspigler** [comment]
This approach doesn't seem as good to me as the other proposed changes.  Why only show a subset of config options as comments? Doesn't really matter if it is a template provided by a third party.

**Comment by prayank23** [comment]
> Why only show a subset of config options as comments? Doesn't really matter if it is a template provided by a third party.

We will have to decide and mention only few options in template. For other options users can do research. This file can also be left blank but one template will help users understand basic options.

As mentioned in OP options can be added/removed based on feedback.

**Comment by Rspigler** [comment]
> It's not possible to add all options

I don't see why that's true.  Look at this PR for example https://github.com/bitcoin/bitcoin/pull/22235

**Comment by prayank23** [comment]
Sorry I don't agree with that approach to solve this problem. I would prefer a template in bitcoin.conf with basic things that a new user could understand and file created on first run.

File: src/conf_file.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# Options only for regtest
+#[regtest]
+
+# Options only for regtest
```
Comment (luke-jr) [inline src/conf_file.h:50]: Incorrect comment

Comment (luke-jr) [inline src/conf_file.h:50]: I don't think regtest/signet should be included here. They're not basic usage

File: src/conf_file.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+# Query for peer addresses via DNS lookup, if low on addresses.
+#dnsseed=0
+# Specify your own public IP address.
+#externalip=your_tor_address.onion
+# Accept incoming connections from peers.
+#listen=0
+# Use separate SOCKS5 proxy <ip:port> to reach peers via Tor hidden services.
+#onion=127.0.0.1:9050
+# Only connect to peers via Tor.
+#onlynet=onion
```
Comment (luke-jr) [inline src/conf_file.h:28]: This is a bad idea, so shouldn't be implicitly recommended

File: src/util/system.cpp (hunk 2 above)
Comment (luke-jr) [reply src/util/system.cpp:451]: But you moved the code the comment was "attached to" into that new function...

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/util/system.cpp (hunk 2 above)
Comment (prayank23) [reply src/util/system.cpp]: Moved in https://github.com/bitcoin/bitcoin/pull/23931/commits/ccbd9a68c00d1588a8846aa43a6f2b3777255393

File: src/conf_file.h (hunk 4 above)
Comment (prayank23) [reply src/conf_file.h]: Removed in https://github.com/bitcoin/bitcoin/pull/23931/commits/8a5f2b972874a6e65678a680c627e8f6bb411a3a

File: src/conf_file.h (hunk 3 above)
Comment (prayank23) [reply src/conf_file.h]: Removed in https://github.com/bitcoin/bitcoin/pull/23931/commits/6c623e4cbdec7099a2ae68ca15a830d6455ba1d5

Comment (prayank23) [reply src/conf_file.h]: Removed these two in https://github.com/bitcoin/bitcoin/pull/23931/commits/6c623e4cbdec7099a2ae68ca15a830d6455ba1d5

File: src/conf_file.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+static const char* DEFAULT_BITCOIN_CONF_TEXT = R"(##
+## bitcoin.conf configuration file
+##
+
+# [network]
+# Add a node IP address to connect to and attempt to keep the connection open. This option can be set multiple times.
+#addnode=randomchars.onion
+# Bind to given address and always listen on it. (default: 0.0.0.0). Use [host]:port notation for IPv6. Append =onion to tag any incoming connections to that address and port as incoming Tor connections
+#bind=127.0.0.1:8333
```
Comment (niVelion) [inline src/conf_file.h:16]: The default behaviour is to listen on 0.0.0.0; to listen on all interfaces.

In my opinion this option doesn't need to be mentioned here - if the aim is to include "basic things that a new user could understand", the default behaviour will suit most users, whereas the implicit suggestion to bind to localhost thereby explicitly preventing remote connections, seems harmful.

Comment (prayank23) [reply src/conf_file.h:16]: > I was referring to L16 #bind=127.0.0.1:833

Its an example for using `bind`

Comment (niVelion) [reply src/conf_file.h:16]: You mention the default binding is 0.0.0.0; which I understand to mean the server will listen on all network interfaces.

And we know listening on 127.0.0.1 would instruct the server not to listen on any interface; not to receive remote connections, but even a technical user might not realize this.

Maybe swapping the default with the example would reduce user error?

```suggestion
 # Bind to given address and always listen on it (eg 192.249.249.1). Use [host]:port notation for IPv6. Append =onion to tag any incoming connections to that address and port as incoming Tor connections 
#bind= 0.0.0.0:8333
```

**Pull request closed without merging**